Doxygen support in CMake

zebmason opened the pull request:
By default it doesn't build the docs and no longer builds LaTeX. Are those two assumptions correct?

See also issue #225

matt-gretton-dann commented on CMakeLists.txt:
```diff
@@ -56,3 +56,31 @@ else()
+    message("Doxygen build started")
+
+    # note the option ALL which allows to build the docs together with the application
+    add_custom_target( doxygen ALL
```
I would prefer this to be a `docs` target rather than `all` (following the principles in https://www.gnu.org/prep/standards/html_node/Standard-Targets.html).

zebmason replied:
I have tried changing to the target to `html` but I cannot then figure out how to build the docs in VS 2019.

matt-gretton-dann replied:
I _think_ what we want is:

```cmake
add_custom_target(doxygen ...) # Without the ALL
add_custom_target(docs DEPENDS doxygen)
```

zebmason replied:
This was convoluted to fix

zebmason replied:
Thanks, I've added an ALL to the second line. First time I've done any high level CMake work in 6 months.

zebmason commented on docs/Doxyfile.in:
```diff
@@ -2394,7 +2396,7 @@ INTERACTIVE_SVG        = NO
 # This tag requires that the tag HAVE_DOT is set to YES.
 
-DOT_PATH               = 
+DOT_PATH               = "@DOXYGEN_DOT_PATH@"
```
This works fine with VS 2019, I haven't fired up an Ubuntu VM in days but I'll double check against the manual

zebmason replied:
Have updated as DOXYGEN_DOT_PATH is deprecated

zebmason commented on docs/Doxyfile.in:
```diff
@@ -435,7 +435,7 @@ LOOKUP_CACHE_SIZE      = 0
 # The default value is: NO.
 
-EXTRACT_ALL            = YES
+EXTRACT_ALL            = NO
```
Otherwise warnings are disabled

zebmason commented on docs/Doxyfile.in:
```diff
@@ -756,7 +756,7 @@ WARN_IF_DOC_ERROR      = YES
 # The default value is: NO.
 
-WARN_NO_PARAMDOC       = NO
+WARN_NO_PARAMDOC       = YES
```
Currently 1017 warnings of which 10 relate to markdown

matt-gretton-dann replied:
What are the markdown warnings?  (Can we have a PR to fix them first?).

zebmason commented on CMakeLists.txt:
```diff
@@ -56,3 +56,31 @@ else()
 endif()
+
+# first we can indicate the documentation build as an option and set it to OFF by default
+option(BUILD_DOC "Build documentation" OFF)
```
Will do

zebmason commented on docs/Doxyfile.in:
```diff
@@ -790,7 +790,9 @@ WARN_LOGFILE           =
 # Note: If this tag is empty the current directory is searched.
 
-INPUT                  = ./src
+INPUT                  = @CMAKE_CURRENT_SOURCE_DIR@/src \
```
doxygen is definitely fine with Unix file paths on Windows here

zebmason commented on CMakeLists.txt:
```diff
@@ -56,3 +56,31 @@ else()
+
+# check if Doxygen is installed
+find_package(Doxygen)
+if (DOXYGEN_FOUND AND BUILD_DOC)
```
Done, I've also changed the tabs to 2 spaces

matt-gretton-dann commented:
If I do:

```sh
make doxygen
make doxygen
```

Will it rebuild the docs from scratch each time or does it only update any changed files?  (We want it to do that latter).

zebmason commented:
It does the docs from scratch each time hence the option to switch it off. Please see the changes in docs/build.md

zebmason commented:
Perhaps should change BUILD_DOC back to off by default?

zebmason commented:
@matt-gretton-dann Just as a matter of interest where does doxygen.yml push the docs to? A public website?

matt-gretton-dann commented:
> @matt-gretton-dann Just as a matter of interest where does doxygen.yml push the docs to? A public website?

In theory yes: https://mrc-ide.github.io/

However, there are some issues with pushing that so its 404ing at the moment.

zebmason commented:
@matt-gretton-dann 
Just set the `BUILD_DOC` CMake option back to `OFF` by default
Updated the yml to do a `make doxygen` rather than the longer `cmake ...`
Updated `build.md` so it hopefully clearer
